src/transform/src/predicate_pushdown.rs

```diff
@@ -785,35 +789,48 @@ impl PredicatePushdown {
                 expr2: eqinnerexpr2,
             } = &mut **expr2
             {
-                if let MirScalarExpr::CallBinary {
-                    func: BinaryFunc::And,
-                    expr1: andinnerexpr1,
-                    expr2: andinnerexpr2,
-                } = &mut **expr1
+                let isnull1 = eqinnerexpr1.clone().call_unary(UnaryFunc::IsNull);
+                let isnull2 = eqinnerexpr2.clone().call_unary(UnaryFunc::IsNull);
+                let both_null = isnull1.call_binary(isnull2, BinaryFunc::And);
+
+                if Self::extract_reduced_conjunction_terms(both_null, relation_type)
+                    == Self::extract_reduced_conjunction_terms((**expr1).clone(), relation_type)
                 {
-                    if let MirScalarExpr::CallUnary {
-                        func: UnaryFunc::IsNull,
-                        expr: nullexpr1,
-                    } = &**andinnerexpr1
-                    {
-                        if let MirScalarExpr::CallUnary {
-                            func: UnaryFunc::IsNull,
-                            expr: nullexpr2,
-                        } = &**andinnerexpr2
-                        {
-                            if (&**eqinnerexpr1 == &**nullexpr1)
-                                && (&**eqinnerexpr2 == &**nullexpr2)
-                            {
-                                return Some(((**eqinnerexpr1).clone(), (**eqinnerexpr2).clone()));
-                            }
-                        }
-                    }
+                    return Some(((**eqinnerexpr1).clone(), (**eqinnerexpr2).clone()));
                 }
             }
         }
         None
     }
 
+    /// Reduces the given expression and returns its AND-ed terms.
+    fn extract_reduced_conjunction_terms(
+        mut s: MirScalarExpr,
+        relation_type: &repr::RelationType,
+    ) -> Vec<MirScalarExpr> {
+        s.reduce(relation_type);
+
+        let mut pending = vec![s];
+        let mut terms = Vec::new();
+
+        while let Some(expr) = pending.pop() {
+            if let MirScalarExpr::CallBinary {
```

Comment:
Can we have a test for `s` where `s` would consist of multiple `and` statements?

Reply:
Sure. I added a few. Although I must admit that case is probably taken care of by some other expression simplication code, since at the end, we are looking for an AND-tree with only two distinct expressions: `isnull(equinnexpr1)` and `isnull(equinnexpr2)`